src/util/ActiveShapeUtils.tsx

```diff
@@ -63,3 +78,140 @@ export function Shape<OptionType, ExtraProps, ShapePropsType>({
+  ? PieItem
+  : DisplayName extends 'Scatter'
+  ? ScatterItem
+  : never;
+
+type GetActiveShapeIndexForTooltip = {
+  itemDisplayName: ItemDisplayName;
+  activeTooltipItem: ShapeData<ItemDisplayName>;
+  graphicalItem: {
+    props: Record<GraphicalItemShapeKey, unknown[]>;
+  };
+  itemData: unknown[];
+};
+
+function isFunnel(itemDisplayName: ItemDisplayName, _item: unknown): _item is FunnelItem {
+  return itemDisplayName === 'Funnel';
+}
+
+function isPie(itemDisplayName: ItemDisplayName, _item: unknown): _item is PieItem {
+  return itemDisplayName === 'Pie';
+}
+
+function isScatter(itemDisplayName: ItemDisplayName, _item: unknown): _item is ScatterItem {
+  return itemDisplayName === 'Scatter';
+}
+
+function getShapeDataKey(itemDisplayName: ItemDisplayName): GraphicalItemShapeKey {
+  let shapeKey: GraphicalItemShapeKey;
+
+  if (itemDisplayName === 'Funnel') {
+    shapeKey = 'trapezoids';
+  } else if (itemDisplayName === 'Pie') {
+    shapeKey = 'sectors';
+  } else if (itemDisplayName === 'Scatter') {
+    shapeKey = 'points';
+  }
+
+  return shapeKey;
+}
+
+/**
+ *
+ * @param {GetActiveShapeIndexForTooltip} arg an object of incoming attributes from Tooltip
+ * @returns {number}
+ *
+ * To handle possible duplicates in the data set,
+ * match both the data value of the active item to a data value on a graph item,
+ * and match the mouse coordinates of the active item to the coordinates of in a particular components shape data.
+ * This assumes equal lengths of shape objects to data items.
+ */
+export function getActiveShapeIndexForTooltip({
+  itemDisplayName,
+  activeTooltipItem,
+  graphicalItem,
+  itemData,
+}: GetActiveShapeIndexForTooltip): number {
+  const shapeKey = getShapeDataKey(itemDisplayName);
+
+  const tooltipPayload: Partial<ShapeData<typeof itemDisplayName>> = isScatter(itemDisplayName, activeTooltipItem)
+    ? activeTooltipItem.payload
+    : activeTooltipItem.tooltipPayload?.[0].payload.payload;
+
+  const activeItemMatches = itemData.filter((datum: any, dataIndex: number) => {
+    const valuesMatch = _.isEqual(tooltipPayload, datum);
+
+    const mouseCoordinateMatches = graphicalItem.props[shapeKey].filter(
+      (shapeData: ShapeData<typeof itemDisplayName>) => {
+        if (isFunnel(itemDisplayName, shapeData) && isFunnel(itemDisplayName, activeTooltipItem)) {
+          const xMatches = shapeData.x === activeTooltipItem?.labelViewBox?.x || shapeData.x === activeTooltipItem.x;
+          const yMatches = shapeData.y === activeTooltipItem?.labelViewBox?.y || shapeData.y === activeTooltipItem.y;
+          return xMatches && yMatches;
+        }
+
+        if (isPie(itemDisplayName, shapeData) && isPie(itemDisplayName, activeTooltipItem)) {
+          const startAngleMatches = shapeData.endAngle === activeTooltipItem.endAngle;
+          const endAngleMatches = shapeData.startAngle === activeTooltipItem.startAngle;
+          return startAngleMatches && endAngleMatches;
+        }
```

Comment:
Could we instead pass a comparison function? That could simplify this component a little.

Reply:
Or, we could simply pass a list of keys for comparison. 
For the Pie: ['endAngle', 'startAngle'], for the scatter: ['x', 'y', 'z'], etc. 

Just wondering whether we can simplify this function to not know about Pies, Scatters, etc.

Reply:
List of keys won't work because I need to do an `||` comparison a few lines above this for funnel.

I can use the type guards with the incoming `graphicalItem` and `activeItem` to get the different comparisons without this function needing to know anything about the component type.